src/pino_ode_solve.jl

```diff
@@ -0,0 +1,327 @@
+* `kwargs`: Extra keyword arguments are splatted to the Optimization.jl `solve` call.
+
+## References
+
+* Sifan Wang "Learning the solution operator of parametric partial differential equations with physics-informed DeepOnets"
+* Zongyi Li "Physics-Informed Neural Operator for Learning Partial Differential Equations"
+"""
+@concrete struct PINOODE
+    chain
+    opt
+    bounds
+    number_of_parameters::Int
+    init_params
+    strategy <: Union{Nothing, AbstractTrainingStrategy}
+    additional_loss <: Union{Nothing, Function}
+    kwargs
+end
+
+function PINOODE(chain,
+        opt,
+        bounds,
+        number_of_parameters;
+        init_params = nothing,
+        strategy = nothing,
+        additional_loss = nothing,
+        kwargs...)
+    chain isa AbstractLuxLayer || (chain = FromFluxAdaptor()(chain))
+    return PINOODE(chain, opt, bounds, number_of_parameters,
+        init_params, strategy, additional_loss, kwargs)
+end
+
+@concrete struct PINOPhi
+    model <: AbstractLuxLayer
+    smodel <: StatefulLuxLayer
+end
+
+function PINOPhi(model::AbstractLuxLayer, st)
+    return PINOPhi(model, StatefulLuxLayer{false}(model, nothing, st))
+end
+
+function generate_pino_phi_θ(chain::AbstractLuxLayer, ::Nothing)
+    θ, st = LuxCore.setup(Random.default_rng(), chain)
+    PINOPhi(chain, st), θ
+end
+
+function generate_pino_phi_θ(chain::AbstractLuxLayer, init_params)
+    st = LuxCore.initialstates(Random.default_rng(), chain)
+    PINOPhi(chain, st), init_params
+end
+
+function (f::PINOPhi{C, T})(x, θ) where {C <: AbstractLuxLayer, T}
+    dev = safe_get_device(θ)
+    return f(dev, safe_expand(dev, x), θ)
+end
+
+function (f::PINOPhi{C, T})(dev, x, θ) where {C <: AbstractLuxLayer, T}
+    f.smodel(dev(x), θ)
+end
+
+function dfdx(phi::PINOPhi{C, T}, x::Tuple, θ) where {C <: DeepONet, T}
+    p, t = x
+    branch_left, branch_right = p, p
+    trunk_left, trunk_right = t .+ sqrt(eps(eltype(t))), t
+    x_left = (branch_left, trunk_left)
+    x_right = (branch_right, trunk_right)
+    (phi(x_left, θ) .- phi(x_right, θ)) ./ sqrt(eps(eltype(t)))
+end
+
+function dfdx(phi::PINOPhi{C, T}, x::Array,
+        θ) where {C <: Lux.Chain, T}
+    ε = [zeros(eltype(x), size(x)[1] - 1)..., sqrt(eps(eltype(x)))]
+    (phi(x .+ ε, θ) - phi(x, θ)) ./ sqrt(eps(eltype(x)))
+end
+
+function physics_loss(
+        phi::PINOPhi{C, T}, prob::ODEProblem, x::Tuple, θ) where {C <: DeepONet, T}
+    p, t = x
+    f = prob.f
+    out = phi(x, θ)
+    if size(p, 1) == 1
+        f_vec = reduce(hcat,
+            [reduce(vcat, [f(out[j, i], p[1, i], t[j]) for j in axes(t, 2)])
+             for i in axes(p, 2)])
+    else
+        f_vec = reduce(hcat,
+            [reduce(vcat, [f(out[j, i], p[:, i], t[j]) for j in axes(t, 2)])
+             for i in axes(p, 2)])
+    end
+    du = dfdx(phi, x, θ)
+    norm = prod(size(du))
+    sum(abs2, du .- f_vec) / norm
+end
```

Comment:
why can't this use the ODE code?

Reply:
Here it has parameters and coordinates on input, unlike the usual ODE with just cord, also ODE does not process DeepONet and NeuralOperator in general. It can't be reduced to instance ODE task. If only on the contrary, ODE is a special case of PinoODE as a parametric ODE with one parameter